compiler-core/src/javascript/expression.rs

```diff
@@ -1411,6 +1302,490 @@ impl<'module, 'a> Generator<'module, 'a> {
+                }
+
+                // UTF8 strings
+                [Opt::Utf8 { .. }] => {
+                    self.tracker.string_bit_array_segment_used = true;
+                    Ok(docvec!["stringBits(", value, ")"])
+                }
+
+                // UTF8 codepoints
+                [Opt::Utf8Codepoint { .. }] => {
+                    self.tracker.codepoint_bit_array_segment_used = true;
+                    Ok(docvec!["codepointBits(", value, ")"])
+                }
+
+                // Bit arrays
+                [Opt::Bits { .. }] => Ok(value),
+
+                // Bit arrays with explicit size. The explicit size slices the bit array to the
+                // specified size. A runtime exception is thrown if the size exceeds the number
+                // of bits in the bit array.
+                [Opt::Bits { .. }, Opt::Size { value: size, .. }]
+                | [Opt::Size { value: size, .. }, Opt::Bits { .. }] => match &**size {
+                    Constant::Int { value: size, .. } => {
+                        self.tracker.bit_array_slice_used = true;
+                        Ok(docvec!["bitArraySlice(", value, ", 0, ", size, ")"])
+                    }
+
+                    _ => Err(Error::Unsupported {
+                        feature: "This bit array segment option".into(),
+                        location: segment.location,
+                    }),
+                },
+
+                // Anything else
+                _ => Err(Error::Unsupported {
+                    feature: "This bit array segment option".into(),
+                    location: segment.location,
+                }),
+            }
+        }))?;
+
+        Ok(docvec!["toBitArray(", segments_array, ")"])
+    }
+
+    fn constant_sized_bit_array_segment_details(
+        &mut self,
+        segment: &'a TypedConstantBitArraySegment,
+    ) -> Result<SizedBitArraySegmentDetails<'a>, Error> {
+        let size = segment.size();
+        let unit = segment.unit();
+        let (size_value, size) = match size {
+            Some(Constant::Int { int_value, .. }) => {
+                let size_value = int_value * unit;
+                let size = eco_format!("{}", size_value).to_doc();
+                (Some(size_value), size)
+            }
+
+            Some(size) => {
+                let mut size = self.constant_expression(Context::Constant, size)?;
+                if unit != 1 {
+                    size = size.group().append(" * ".to_doc().append(unit.to_doc()));
+                }
+
+                (None, size)
+            }
+
+            None => {
+                let size_value: usize = if segment.type_.is_int() { 8 } else { 64 };
+                (Some(BigInt::from(size_value)), docvec![size_value])
+            }
+        };
+
+        Ok(SizedBitArraySegmentDetails { size, size_value })
+    }
+
+    pub(crate) fn guard(&mut self, guard: &'a TypedClauseGuard) -> Output<'a> {
```

Comment:
All the code here up to line ~1900 is basically moved as-is from the pattern module where it was previously